### PR TITLE
Automatically assign Basic Tier to Identities 

### PR DESCRIPTION
**Pull request by Dannyps**
# Readiness checklist

- [x] I added/updated unit tests.
- [ ] I added/updated integration tests.
- [x] I ensured that the PR title is good enough for the changelog.
- [x] I labeled the PR.

### CODE REVIEW

**Review by tnotheis — CHANGES_REQUESTED** [review]
Can you generate a unique database index for `Tier.Name` please? See https://github.com/nmshd/backbone/blob/main/Modules/Synchronization/src/Synchronization.Infrastructure/Persistence/Database/Configurations/DatawalletModificationEntityTypeConfiguration.cs#L11-L12 for an example

File: Modules/Devices/src/Devices.Domain/Entities/Identity.cs [hunk 1]
```diff
@@ -25,6 +27,8 @@ public Identity(string? clientId, IdentityAddress address, byte[] publicKey, byt
 
     public byte IdentityVersion { get; set; }
 
+    public Tier? Tier { get; set; }
```
[the remark this answers is not in the document]
Comment (Dannyps) [reply Modules/Devices/src/Devices.Domain/Entities/Identity.cs:30]: @tnotheis We have a chicken and egg problem here. The database may already have Identities, which do not have a Tier. When we add a TierId to the Identities table (via Migrations), we must set that collumn as nullable, hence that code. 
The alternative - having a default value - is inviable because the Basic Tier may not exist yet.
How should we address this?

[the remark this answers is not in the document]
Comment (Dannyps) [reply Modules/Devices/src/Devices.Domain/Entities/Identity.cs:30]: A possible solution seems to be to make the attribute mandatory but tweak the migration to make the column nullable. I do not know how that can impact the runtime of the application though.

[the remark this answers is not in the document]
Comment (tnotheis) [reply Modules/Devices/src/Devices.Domain/Entities/Identity.cs:30]: Okay, let's try to go with your suggested approach: make the property mandatory and make the column nullable. But don't edit the migration directly. Instead, you should edit the `IdentityEntityTypeConfiguration`:

```c#
builder.Property(x=>x.Tier).IsRequired(false)
```
I'm not sure though if that works in case of a navigation property. If not, I would suggest to leave it as it is for now, and try to remember to fix it in the future when the possibility that there are still people running a version without Tiers is almost 0.

[the remark this answers is not in the document]
Comment (Dannyps) [reply Modules/Devices/src/Devices.Domain/Entities/Identity.cs:30]: We're using `.HasOne()`, which does not support `IsRequired` nor anything similar.

This SO answer seems to shed some light on the issue at hand: https://stackoverflow.com/a/35562728

[the remark this answers is not in the document]
Comment (tnotheis) [reply Modules/Devices/src/Devices.Domain/Entities/Identity.cs:30]: Yeah, okay. Let's just ignore that for now. It's too much pain for too less benefit.

File: Modules/Devices/src/Devices.Application/Identities/Commands/CreateIdentity/Handler.cs [hunk 2]
```diff
@@ -44,7 +45,9 @@ public async Task<CreateIdentityResponse> Handle(CreateIdentityCommand command,
         if (existingIdentity != null)
             throw new OperationFailedException(ApplicationErrors.Devices.AddressAlreadyExists());
 
-        var newIdentity = new Identity(command.ClientId, address, command.IdentityPublicKey, command.IdentityVersion);
+        var basicTier = await _dbContext.Set<Tier>().BasicTier(cancellationToken);
```
[the remark this answers is not in the document]
Comment (Dannyps) [reply Modules/Devices/src/Devices.Application/Identities/Commands/CreateIdentity/Handler.cs:48]: I have done this, but I'm not happy about the solution. See 02c60d2294debef148884b5519bfbb758564fde2

[the remark this answers is not in the document]
Comment (tnotheis) [reply Modules/Devices/src/Devices.Application/Identities/Commands/CreateIdentity/Handler.cs]: Why not? I don't see a problem with it.

File: Modules/Devices/src/Devices.Domain/Entities/Identity.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (tnotheis) [reply Modules/Devices/src/Devices.Domain/Entities/Identity.cs:29]: No worries. I would have done it myself. 😆

File: Modules/Devices/src/Devices.Application/Identities/Commands/CreateIdentity/Handler.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Dannyps) [reply Modules/Devices/src/Devices.Application/Identities/Commands/CreateIdentity/Handler.cs]: It feels that we just added redundant code, but I understand the logic. Ok, cool

**Review by Dannyps — COMMENTED** [review]
The requested changes have been done.

File: Backbone.API/DevicesDbContextSeed.cs [hunk 3]
```diff
@@ -32,15 +32,20 @@ private async Task SeedApplicationUsers(DevicesDbContext context)
         if (await context.Users.AnyAsync())
             return;
 
+        var basicTier = await context.Tiers.GetBasicTier(CancellationToken.None);
+        if (basicTier == null)
+            return;
```
Comment (tnotheis) [inline Backbone.API/DevicesDbContextSeed.cs:37]: You should throw an exception if it doesn't exist, so that we know that something's wrong. You can stick with the one you threw in the previous version:
```c#
throw new Exception("Basic Tier was not found.");
```

File: Backbone.API/DevicesDbContextSeed.cs [hunk 4]
```diff
@@ -75,20 +81,16 @@ private async Task SeedBasicTier(DevicesDbContext context)
...
+
+        var identitiesWithoutTier = context.Identities.Where(i => i.Tier == null).ToList();
+        identitiesWithoutTier.ForEach(it =>
         {
-            throw new Exception("Basic Tier was not found.");
-        }
+            it.Tier = basicTier;
+        });
+
+        await context.SaveChangesAsync();
```
Comment (tnotheis) [inline Backbone.API/DevicesDbContextSeed.cs]: You should replace it with

```c#
await context.Identities.Where(i => i.Tier == null).ExecuteUpdateAsync(s => s.SetProperty(i => i.Tier, basicTier));
```
in order to reduce the number of database roundtrips. Depending on the number of Identities, this could make a huge difference.

I never used the `ExecuteUpdateAsync` method before, so I don't know if it works with navigation properties.

File: Modules/Devices/src/Devices.Domain/Entities/Identity.cs [hunk 5]
```diff
@@ -6,15 +6,14 @@ namespace Backbone.Modules.Devices.Domain.Entities;
 
 public class Identity
 {
-    public Identity(string? clientId, IdentityAddress address, byte[] publicKey, Tier tier, byte identityVersion)
+    public Identity(string? clientId, IdentityAddress address, byte[] publicKey, byte identityVersion)
```
Comment (tnotheis) [inline Modules/Devices/src/Devices.Domain/Entities/Identity.cs]: Why did you remove the Tier parameter from the constructor? 
It shouldn't be possible to create an Identity without a Tier.

File: Modules/Devices/src/Devices.Infrastructure.Database.SqlServer/Migrations/20230413132800_IdentityBasicTier.cs [hunk 6]
```diff
@@ -29,7 +36,8 @@ protected override void Up(MigrationBuilder migrationBuilder)
                 table: "Identities",
                 column: "TierId",
                 principalTable: "Tier",
-                principalColumn: "Id");
+                principalColumn: "Id",
+                onDelete: ReferentialAction.Cascade);
```
Comment (tnotheis) [inline Modules/Devices/src/Devices.Infrastructure.Database.SqlServer/Migrations/20230413132800_IdentityBasicTier.cs:40]: This is evil. This means that if we delete a Tier, all Identities assigned to the Tier will be deleted.

**Review by tnotheis — CHANGES_REQUESTED** [review]

File: Backbone.API/DevicesDbContextSeed.cs (hunk 4 above)
Comment (Dannyps) [reply Backbone.API/DevicesDbContextSeed.cs]: I didn't pay a lot of attention to this because I supposed all changes would be commited at once on `await context.SaveChangesAsync();`.

It seems EF cannot handle it when we set a new navigation property. I'm trying to workaround the issue.

File: Modules/Devices/src/Devices.Domain/Entities/Identity.cs (hunk 5 above)
Comment (Dannyps) [reply Modules/Devices/src/Devices.Domain/Entities/Identity.cs]: EF needs to create instances of the object and cannot have a navigation property in the constructuor, as far as I could see. The error I got was the following:

```
System.InvalidOperationException: 'No suitable constructor was found for entity type 'Identity'. The following constructors had parameters that could not be bound to properties of the entity type: cannot bind 'tier' in 'Identity(..., Tier tier)'.'
```

File: Modules/Devices/src/Devices.Infrastructure.Database.SqlServer/Migrations/20230413132800_IdentityBasicTier.cs (hunk 6 above)
Comment (Dannyps) [reply Modules/Devices/src/Devices.Infrastructure.Database.SqlServer/Migrations/20230413132800_IdentityBasicTier.cs:40]: Indeed! Should we set those as Null, or restrict the delete? We do not have a default value.

File: Backbone.API/DevicesDbContextSeed.cs (hunk 4 above)
Comment (tnotheis) [reply Backbone.API/DevicesDbContextSeed.cs]: > I supposed all changes would be commited at once on `await context.SaveChangesAsync();`.

They are commited at once. But you still need 2 roundtrips: one for the update and one for the select (line 88).

If my suggested solution doesn't work for navigation properties, you should at least call `ToListAsync` instead of `ToList` in line 88.

File: Modules/Devices/src/Devices.Domain/Entities/Identity.cs (hunk 5 above)
Comment (tnotheis) [reply Modules/Devices/src/Devices.Domain/Entities/Identity.cs]: In that case, we should change it completely (I wanted to suggest that already a few days ago): instead of saving a reference to the Tier, we should only save the TierId. That's the solution DDD suggests anyway (when referencing other Aggregates, only do so via their Id and never the whole object)

File: Modules/Devices/src/Devices.Infrastructure.Database.SqlServer/Migrations/20230413132800_IdentityBasicTier.cs (hunk 6 above)
Comment (tnotheis) [reply Modules/Devices/src/Devices.Infrastructure.Database.SqlServer/Migrations/20230413132800_IdentityBasicTier.cs:40]: Since we don't want to have Identities without a Tier, we should go with Restrict.

File: Modules/Devices/src/Devices.Domain/Entities/Identity.cs (hunk 5 above)
Comment (Dannyps) [reply Modules/Devices/src/Devices.Domain/Entities/Identity.cs]: In this case, will we have both a `Tier` and a `TierId` in the `Identity`?

Comment (tnotheis) [reply Modules/Devices/src/Devices.Domain/Entities/Identity.cs]: Nope, only the TierId

Comment (Dannyps) [reply Modules/Devices/src/Devices.Domain/Entities/Identity.cs]: The generated migration does not include a ForeignKey. Should I add it manually, or is there a way to specify that `TierId` must be a valid `Tier.Id`. I looked online but couldn't find anything.

File: Modules/Devices/src/Devices.Domain/Aggregates/Tier/TierId.cs [hunk 7]
```diff
@@ -28,4 +28,12 @@ public static TierId Generate()
             return Result.Failure<TierId, DomainError>(validationResult);
         return Result.Success<TierId, DomainError>(new TierId(value));
     }
+
+    public static TierId New()
+    {
+
```
Comment (tnotheis) [inline Modules/Devices/src/Devices.Domain/Aggregates/Tier/TierId.cs:34]: Redundant blank line

File: Modules/Devices/src/Devices.Domain/Aggregates/Tier/TierId.cs [hunk 8]
```diff
@@ -28,4 +28,12 @@ public static TierId Generate()
             return Result.Failure<TierId, DomainError>(validationResult);
         return Result.Success<TierId, DomainError>(new TierId(value));
     }
+
+    public static TierId New()
+    {
+   
+        var tierIdAsString = StringUtils.Generate(DefaultValidChars, DEFAULT_MAX_LENGTH_WITHOUT_PREFIX);
```
Comment (tnotheis) [inline Modules/Devices/src/Devices.Domain/Aggregates/Tier/TierId.cs:35]:
```suggestion
        var randomPart = StringUtils.Generate(DefaultValidChars, DEFAULT_MAX_LENGTH_WITHOUT_PREFIX);
```

File: Modules/Devices/src/Devices.Domain/Entities/Identity.cs [hunk 9]
```diff
@@ -26,7 +28,8 @@ public Identity(string? clientId, IdentityAddress address, byte[] publicKey, byt
 
     public byte IdentityVersion { get; set; }
 
-    public Tier? Tier { get; set; }
+    [ForeignKey(nameof(Tier))]
```
Comment (tnotheis) [inline Modules/Devices/src/Devices.Domain/Entities/Identity.cs:31]: Remember to remove this

**Review by tnotheis — CHANGES_REQUESTED** [review]

**Review by tnotheis — APPROVED** [review]
